File: leetcode/python/498.diagonal-traverse.py
class Solution:
    def findDiagonalOrder(self, matrix: List[List[int]]) -> List[int]:
        # 对角线和递增

        if not len(matrix) or not len(matrix[0]):
            return []

        res = []
        _reverse = False
        for i in range(len(matrix) + len(matrix[0]) - 1):
            t = []
            beg = 0 if i < len(matrix) else i - len(matrix) + 1
            while (beg <= i and beg < len(matrix[0])):
                t.append(matrix[i - beg][beg])
                beg += 1
            res += t[::-1] if _reverse else t
            _reverse = not _reverse

        return res

File: leetcode/csharp/498.diagonal-traverse.cs
public class Solution
{
    public int[] FindDiagonalOrder(int[][] matrix)
    {
        if (matrix.Length == 0 || matrix[0].Length == 0) return Array.Empty<int>();
        List<List<int>> m = new List<List<int>>();
        for (int i = 0; i < matrix.Length; i++)
        {
            for (int j = 0; j < matrix[i].Length; j++)
            {
                if (i + j >= m.Count)
                {
                    m.Add(new List<int>());
                }

                m[i + j].Add(matrix[i][j]);
            }
        }

        int[] res = new int[matrix.Length * matrix[0].Length];

        int idx = 0;
        bool even = true;
        foreach (var item in m)
        {
            if (even)
            {
                item.Reverse();
            }

            foreach (int v in item)
            {
                res[idx++] = v;
            }

            even = !even;
        }

        return res;
    }
}